File: backend/app/schemas/revision.py
"""
Revision Pydantic schemas
"""
from typing import Optional
from datetime import date, datetime
from uuid import UUID
from pydantic import BaseModel, Field, field_validator


class RevisionBase(BaseModel):
    """Base revision schema"""
    target_article_id: str = Field(..., min_length=1, max_length=100)
    # target_article_pk: removed field
    reason: str = Field(..., min_length=1)
    
    # After fields (all optional)
    after_title: Optional[str] = None
    after_info_category: Optional[UUID] = None  # Changed to UUID type
    after_keywords: Optional[str] = None
    after_importance: Optional[bool] = None
    after_publish_start: Optional[date] = None
    after_publish_end: Optional[date] = None
    after_target: Optional[str] = Field(None, max_length=100)
    after_question: Optional[str] = None
    after_answer: Optional[str] = None
    after_additional_comment: Optional[str] = None


class RevisionCreate(RevisionBase):
    """Schema for creating revisions"""
    approver_id: Optional[UUID] = None
    
    @field_validator('after_publish_start', 'after_publish_end', mode='before')
    @classmethod
    def parse_empty_date(cls, v):
        """Convert empty string to None for optional date fields"""
        if v == "":
            return None
        return v
    
    @field_validator('approver_id', mode='before')
    @classmethod
    def parse_empty_uuid(cls, v):
        """Convert empty string to None for optional UUID fields"""
        if v == "":
            return None
        return v


class RevisionUpdate(BaseModel):
    """Schema for updating revisions"""
    reason: Optional[str] = Field(None, min_length=1)
    status: Optional[str] = Field(None, pattern="^(draft|submitted|approved|rejected|deleted)$")
    processed_at: Optional[datetime] = None  # Add processed_at field
    approver_id: Optional[UUID] = None  # Add approver_id field
    
    # After fields (all optional)
    after_title: Optional[str] = None
    after_info_category: Optional[UUID] = None  # Changed to UUID type
    after_keywords: Optional[str] = None
    after_importance: Optional[bool] = None
    after_publish_start: Optional[date] = None
    after_publish_end: Optional[date] = None
    after_target: Optional[str] = Field(None, max_length=100)
    after_question: Optional[str] = None
    after_answer: Optional[str] = None
    after_additional_comment: Optional[str] = None


class RevisionStatusUpdate(BaseModel):
    """Schema for updating revision status"""
    status: str = Field(..., pattern="^(submitted|approved|rejected|deleted)$")


class RevisionInDB(RevisionBase):
    """Schema for revision in database"""
    revision_id: UUID
    proposer_id: UUID
    status: str
    approver_id: Optional[UUID] = None
    processed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class Revision(RevisionInDB):
    """Schema for revision response"""
    pass


class RevisionWithNames(BaseModel):
    """Schema for revision response with user names instead of IDs, and article number"""
    revision_id: UUID
    article_number: Optional[str] = None  # Article number from Article table
    reason: str
    after_title: Optional[str] = None
    after_info_category: Optional[UUID] = None
    after_keywords: Optional[str] = None
    after_importance: Optional[bool] = None
    after_publish_start: Optional[date] = None
    after_publish_end: Optional[date] = None
    after_target: Optional[str] = Field(None, max_length=100)
    after_question: Optional[str] = None
    after_answer: Optional[str] = None
    after_additional_comment: Optional[str] = None
    status: str
    processed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    
    # User names instead of IDs
    proposer_name: str
    approver_name: Optional[str] = None
    
    class Config:
        from_attributes = True

File: frontend/src/types/revision.ts
/**
 * 修正案関連の型定義
 */

export type RevisionStatus = 'draft' | 'submitted' | 'approved' | 'rejected' | 'deleted';

export interface Revision {
  revision_id: string;  // UUID
  target_article_id: string;
  proposer_id: string;  // UUID
  approver_id: string;  // UUID (required)
  
  // After-only fields (all nullable)
  after_title?: string | null;
  after_info_category?: string | null;  // UUID
  after_keywords?: string | null;
  after_importance?: boolean | null;
  after_publish_start?: string | null;  // ISO 8601 date
  after_publish_end?: string | null;    // ISO 8601 date
  after_target?: string | null;
  after_question?: string | null;
  after_answer?: string | null;
  after_additional_comment?: string | null;
  
  // Metadata
  reason: string;
  status: RevisionStatus;
  processed_at?: string | null;  // ISO 8601 datetime
  created_at: string;  // ISO 8601 datetime
  updated_at: string;  // ISO 8601 datetime
  
  // Names populated from relations (in list views)
  proposer_name?: string;
  approver_name?: string;
  
  // Relations (populated in some endpoints)
  proposer?: unknown;  // User object
  approver?: unknown;  // User object
  article?: unknown;   // Article object
}

export interface RevisionCreate {
  target_article_id: string;
  approver_id: string;  // Required
  reason: string;
  
  // At least one after_* field must be provided
  after_title?: string | null;
  after_info_category?: string | null;
  after_keywords?: string | null;
  after_importance?: boolean | null;
  after_publish_start?: string | null;
  after_publish_end?: string | null;
  after_target?: string | null;
  after_question?: string | null;
  after_answer?: string | null;
  after_additional_comment?: string | null;
}

export interface RevisionUpdate {
  approver_id?: string;
  reason?: string;
  after_title?: string | null;
  after_info_category?: string | null;
  after_keywords?: string | null;
  after_importance?: boolean | null;
  after_publish_start?: string | null;
  after_publish_end?: string | null;
  after_target?: string | null;
  after_question?: string | null;
  after_answer?: string | null;
  after_additional_comment?: string | null;
}

export interface ApprovalDecision {
  action: 'approve' | 'reject' | 'request_changes' | 'defer';
  comment: string;
  priority?: 'low' | 'medium' | 'high' | 'urgent';
}

export interface RevisionStatistics {
  total_count: number;
  by_status: {
    draft: number;
    submitted: number;
    approved: number;
    rejected: number;
    deleted: number;
  };
  recent_submissions: number;
  approval_rate: number;
  average_processing_time_hours?: number;
}